feature/session/src/main/java/io/github/droidkaigi/confsched2019/session/ui/widget/TimetableLayoutManager.kt

```diff
@@ -64,35 +93,57 @@ class TimeTableLayoutManager(
     private var firstStartUnixMin = NO_TIME
     private var lastEndUnixMin = NO_TIME
 
+    private var saveState: SaveState? = null
+
     override fun generateDefaultLayoutParams(): RecyclerView.LayoutParams {
         return RecyclerView.LayoutParams(
             RecyclerView.LayoutParams.WRAP_CONTENT,
             RecyclerView.LayoutParams.WRAP_CONTENT
         )
     }
 
+    override fun onRestoreInstanceState(state: Parcelable?) {
+        saveState = (state as? SaveState)
+    }
+
+    override fun onSaveInstanceState(): Parcelable? {
+        if (childCount == 0) return null
+
+        val view = findFirstVisibleView() ?: return null
+        return SaveState(
+            view.adapterPosition,
+            getDecoratedLeft(view),
+            getDecoratedTop(view)
+        )
+    }
+
     override fun onLayoutChildren(recycler: Recycler, state: State) {
         if (itemCount == 0) {
             detachAndScrapAttachedViews(recycler)
             periods.clear()
             columns.clear()
             anchor.reset()
+            saveState = null
             return
         }
 
-        detachAndScrapAttachedViews(recycler)
         anchor.reset()
         calculateColumns()
 
-        var offsetX = parentLeft
-        for (columnNum in 0 until columns.size()) {
-            if (columns.indexOfKey(columnNum) == 0) anchor.leftColumn = columnNum
-            offsetX += fillColumnHorizontally(columnNum, 0, offsetX, parentTop, true, recycler)
-            if (offsetX > parentRight) {
-                anchor.rightColumn = columnNum
-                break
-            }
-        }
+        val firstVisibleView = if (childCount > 0) findFirstVisibleView() else null
+        val offsetX = saveState?.left ?: firstVisibleView?.let(this::getDecoratedLeft)
+        val offsetY = saveState?.top ?: firstVisibleView?.let(this::getDecoratedTop)
+        val period = saveState?.position?.let(periods::getOrNull)
```

Comment:
Can we simplify like this?

`(saveState?.position ?: firstVisibleView?.adapterPosition)?.let(periods::getOrNull)`

I'm afraid `periods::getOrNull` can return null as well. Which of mine or yours is correct?